#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
KETI Analysis Engine gRPC Client Test
gRPC ì„œë²„ í…ŒìŠ¤íŠ¸ìš© í´ë¼ì´ì–¸íŠ¸
"""

import grpc
import analysis_service_pb2 as pb2
import analysis_service_pb2_grpc as pb2_grpc
import logging
import time

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class AnalysisClient:
    def __init__(self, host='localhost', port=50051):
        self.host = host
        self.port = port
        self.channel = None
        self.stub = None
    
    def connect(self):
        """gRPC ì„œë²„ì— ì—°ê²°"""
        try:
            self.channel = grpc.insecure_channel(f'{self.host}:{self.port}')
            self.stub = pb2_grpc.AnalysisServiceStub(self.channel)
            logger.info(f"gRPC ì„œë²„ì— ì—°ê²°: {self.host}:{self.port}")
            return True
        except Exception as e:
            logger.error(f"ì—°ê²° ì‹¤íŒ¨: {e}")
            return False
    
    def test_create_turtlebot(self, device_id="test-bot-01", model="TURTLEBOT3-Burger", location="Lab-A"):
        """í„°í‹€ë´‡ ìƒì„± í…ŒìŠ¤íŠ¸"""
        try:
            request = pb2.CreateTurtlebotRequest(
                device_id=device_id,
                model=model,
                location=location
            )
            
            response = self.stub.CreateTurtlebot(request)
            logger.info(f"í„°í‹€ë´‡ ìƒì„± ê²°ê³¼: {response.success}, {response.message}")
            
            if response.success and response.device_info:
                logger.info(f"ë””ë°”ì´ìŠ¤ ì •ë³´: ID={response.device_info.device_id}, "
                          f"íƒ€ì…={response.device_info.device_type}, "
                          f"ëª¨ë¸={response.device_info.model}")
            
            return response.success
        except Exception as e:
            logger.error(f"í„°í‹€ë´‡ ìƒì„± ì‹¤íŒ¨: {e}")
            return False
    
    def test_get_device_status(self, device_id="test-bot-01"):
        """ë””ë°”ì´ìŠ¤ ìƒíƒœ ì¡°íšŒ í…ŒìŠ¤íŠ¸"""
        try:
            request = pb2.GetDeviceStatusRequest(device_id=device_id)
            response = self.stub.GetDeviceStatus(request)
            
            logger.info(f"ë””ë°”ì´ìŠ¤ ìƒíƒœ ì¡°íšŒ ê²°ê³¼: {response.success}, {response.message}")
            
            if response.success and response.status:
                logger.info(f"ìƒíƒœ: {response.status.status}, "
                          f"ë°°í„°ë¦¬ ë ˆë²¨: {response.status.battery_level}%, "
                          f"ë°°í„°ë¦¬ Wh: {response.status.battery_wh}")
            
            return response.success
        except Exception as e:
            logger.error(f"ìƒíƒœ ì¡°íšŒ ì‹¤íŒ¨: {e}")
            return False
    
    def test_update_from_influx(self, device_id="TURTLEBOT3-Burger-1"):
        """InfluxDB ì—…ë°ì´íŠ¸ í…ŒìŠ¤íŠ¸"""
        try:
            request = pb2.UpdateFromInfluxRequest(
                device_id=device_id,
                lookback="-30m"
            )
            
            response = self.stub.UpdateFromInflux(request)
            logger.info(f"InfluxDB ì—…ë°ì´íŠ¸ ê²°ê³¼: {response.success}, {response.message}")
            
            if response.success and response.updated_status:
                logger.info(f"ì—…ë°ì´íŠ¸ëœ ìƒíƒœ: {response.updated_status.status}, "
                          f"ë°°í„°ë¦¬ Wh: {response.updated_status.battery_wh}")
            
            return response.success
        except Exception as e:
            logger.error(f"InfluxDB ì—…ë°ì´íŠ¸ ì‹¤íŒ¨: {e}")
            return False
    
    def test_analyze_device(self, device_id="test-bot-01"):
        """ë””ë°”ì´ìŠ¤ ë¶„ì„ í…ŒìŠ¤íŠ¸"""
        try:
            request = pb2.AnalyzeDeviceRequest(device_id=device_id)
            response = self.stub.AnalyzeDevice(request)
            
            logger.info(f"ë””ë°”ì´ìŠ¤ ë¶„ì„ ê²°ê³¼: {response.success}, {response.message}")
            
            if response.success and response.analysis:
                analysis = response.analysis
                logger.info(f"ì„±ëŠ¥ ì ìˆ˜: {analysis.performance_score}")
                logger.info(f"íš¨ìœ¨ì„± ë“±ê¸‰: {analysis.efficiency_rating}")
                logger.info(f"ë°°í„°ë¦¬ ê±´ê°•ë„: {analysis.battery_health}")
                logger.info(f"ë¶„ì„ ìš”ì•½: {analysis.analysis_summary}")
                
                if analysis.metrics:
                    logger.info("ì„¸ë¶€ ë©”íŠ¸ë¦­:")
                    for metric in analysis.metrics:
                        logger.info(f"  - {metric.metric_name}: {metric.value} {metric.unit}")
            
            return response.success
        except Exception as e:
            logger.error(f"ë””ë°”ì´ìŠ¤ ë¶„ì„ ì‹¤íŒ¨: {e}")
            return False
    
    def test_get_all_devices(self):
        """ëª¨ë“  ë””ë°”ì´ìŠ¤ ì¡°íšŒ í…ŒìŠ¤íŠ¸"""
        try:
            request = pb2.GetAllDevicesRequest()
            response = self.stub.GetAllDevices(request)
            
            logger.info(f"ë””ë°”ì´ìŠ¤ ëª©ë¡ ì¡°íšŒ ê²°ê³¼: {response.success}, {response.message}")
            
            if response.success and response.devices:
                logger.info(f"ì´ {len(response.devices)}ê°œì˜ ë””ë°”ì´ìŠ¤:")
                for device in response.devices:
                    logger.info(f"  - {device.device_id} ({device.device_type}): {device.status}")
            
            return response.success
        except Exception as e:
            logger.error(f"ë””ë°”ì´ìŠ¤ ëª©ë¡ ì¡°íšŒ ì‹¤íŒ¨: {e}")
            return False
    
    def test_get_fleet_analysis(self):
        """í”Œë¦¿ ë¶„ì„ í…ŒìŠ¤íŠ¸"""
        try:
            request = pb2.GetFleetAnalysisRequest()
            response = self.stub.GetFleetAnalysis(request)
            
            logger.info(f"í”Œë¦¿ ë¶„ì„ ê²°ê³¼: {response.success}, {response.message}")
            
            if response.success and response.fleet_analysis:
                fleet = response.fleet_analysis
                logger.info(f"ì´ ë””ë°”ì´ìŠ¤: {fleet.total_devices}")
                logger.info(f"í™œì„± ë””ë°”ì´ìŠ¤: {fleet.active_devices}")
                logger.info(f"í‰ê·  ì„±ëŠ¥: {fleet.average_performance}")
                logger.info(f"í‰ê·  ë°°í„°ë¦¬ ê±´ê°•ë„: {fleet.average_battery_health}")
            
            return response.success
        except Exception as e:
            logger.error(f"í”Œë¦¿ ë¶„ì„ ì‹¤íŒ¨: {e}")
            return False
    
    def test_male_mission_analysis(self, device_id="TURTLEBOT3-Burger-1"):
        """MALE Mission ë¶„ì„ í…ŒìŠ¤íŠ¸"""
        try:
            request = pb2.AnalyzeMaleMissionRequest(
                device_id=device_id,
                mission_type="patrol",
                time_range="-24h"
            )
            
            response = self.stub.AnalyzeMaleMission(request)
            logger.info(f"ğŸ¯ MALE Mission ë¶„ì„ ê²°ê³¼: {response.success}, {response.message}")
            
            if response.success and response.analysis:
                analysis = response.analysis
                logger.info(f"ë¯¸ì…˜ ì„±ê³µë¥ : {analysis.mission_success_rate:.1f}%")
                logger.info(f"ë¯¸ì…˜ íš¨ê³¼ì„±: {analysis.mission_effectiveness:.1f}%")
                logger.info(f"í‰ê·  ìˆ˜í–‰ì‹œê°„: {analysis.average_mission_duration:.1f}ë¶„")
                logger.info(f"ì´ ë¯¸ì…˜ ìˆ˜: {len(analysis.mission_records)}")
            
            return response.success
        except Exception as e:
            logger.error(f"MALE Mission ë¶„ì„ ì‹¤íŒ¨: {e}")
            return False
    
    def test_accuracy_analysis(self, device_id="TURTLEBOT3-Burger-1"):
        """ì •í™•ë„ ë¶„ì„ í…ŒìŠ¤íŠ¸"""
        try:
            request = pb2.AnalyzeAccuracyRequest(
                device_id=device_id,
                accuracy_type="positioning",
                time_range="-24h"
            )
            
            response = self.stub.AnalyzeAccuracy(request)
            logger.info(f"ğŸ“ ì •í™•ë„ ë¶„ì„ ê²°ê³¼: {response.success}, {response.message}")
            
            if response.success and response.analysis:
                analysis = response.analysis
                logger.info(f"ì •í™•ë„: {analysis.accuracy_percentage:.1f}%")
                logger.info(f"í‰ê·  ì˜¤ì°¨: {analysis.average_error_distance:.2f}m")
                logger.info(f"ìµœëŒ€ ì˜¤ì°¨: {analysis.max_error_distance:.2f}m")
                logger.info(f"ì¸¡ì • ìˆ˜: {len(analysis.accuracy_records)}")
            
            return response.success
        except Exception as e:
            logger.error(f"ì •í™•ë„ ë¶„ì„ ì‹¤íŒ¨: {e}")
            return False
    
    def test_latency_analysis(self, device_id="TURTLEBOT3-Burger-1"):
        """ì§€ì—°ì‹œê°„ ë¶„ì„ í…ŒìŠ¤íŠ¸"""
        try:
            request = pb2.AnalyzeLatencyRequest(
                device_id=device_id,
                latency_type="command",
                time_range="-24h"
            )
            
            response = self.stub.AnalyzeLatency(request)
            logger.info(f"âš¡ ì§€ì—°ì‹œê°„ ë¶„ì„ ê²°ê³¼: {response.success}, {response.message}")
            
            if response.success and response.analysis:
                analysis = response.analysis
                logger.info(f"í‰ê·  ì§€ì—°ì‹œê°„: {analysis.average_latency_ms:.1f}ms")
                logger.info(f"ìµœëŒ€ ì§€ì—°ì‹œê°„: {analysis.max_latency_ms:.1f}ms")
                logger.info(f"ìµœì†Œ ì§€ì—°ì‹œê°„: {analysis.min_latency_ms:.1f}ms")
                logger.info(f"ì¸¡ì • ìˆ˜: {len(analysis.latency_records)}")
            
            return response.success
        except Exception as e:
            logger.error(f"ì§€ì—°ì‹œê°„ ë¶„ì„ ì‹¤íŒ¨: {e}")
            return False
    
    def test_device_score(self, device_id="TURTLEBOT3-Burger-1"):
        """ë””ë°”ì´ìŠ¤ ì¢…í•© ì ìˆ˜ í…ŒìŠ¤íŠ¸"""
        try:
            request = pb2.GetDeviceScoreRequest(
                device_id=device_id,
                time_range="-24h"
            )
            
            response = self.stub.GetDeviceScore(request)
            logger.info(f"ğŸ† ì¢…í•© ì ìˆ˜ ê²°ê³¼: {response.success}, {response.message}")
            
            if response.success and response.score:
                score = response.score
                logger.info(f"ğŸ“Š ì¢…í•© ì ìˆ˜: {score.overall_score:.1f}/100 (ë“±ê¸‰: {score.grade})")
                logger.info(f"   ì„±ëŠ¥: {score.performance_score:.1f}, ë¯¸ì…˜: {score.mission_score:.1f}")
                logger.info(f"   ì •í™•ë„: {score.accuracy_score:.1f}, ì§€ì—°ì‹œê°„: {score.latency_score:.1f}")
                logger.info(f"   ì‹ ë¢°ì„±: {score.reliability_score:.1f}")
                
                if score.score_details:
                    logger.info("ğŸ“‹ ìƒì„¸ ë¶„ì„:")
                    for detail in score.score_details[:3]:  # ì²˜ìŒ 3ê°œë§Œ ì¶œë ¥
                        logger.info(f"   â€¢ {detail.category}: {detail.score:.1f}")
                        if detail.recommendations:
                            logger.info(f"     ğŸ’¡ ê°œì„ ì•ˆ: {detail.recommendations[0]}")
            
            return response.success
        except Exception as e:
            logger.error(f"ì¢…í•© ì ìˆ˜ ì‹¤íŒ¨: {e}")
            return False
    
    def test_battery_status(self, device_id="TURTLEBOT3-Burger-1"):
        """ë°°í„°ë¦¬ ìƒíƒœ í…ŒìŠ¤íŠ¸"""
        try:
            request = pb2.GetBatteryStatusRequest(device_id=device_id)
            response = self.stub.GetBatteryStatus(request)
            
            logger.info(f"ğŸ”‹ ë°°í„°ë¦¬ ìƒíƒœ ê²°ê³¼: {response.success}, {response.message}")
            
            if response.success and response.battery_status:
                battery = response.battery_status
                logger.info(f"ë°°í„°ë¦¬ ë ˆë²¨: {battery.battery_level:.1f}%")
                logger.info(f"ë°°í„°ë¦¬ Wh: {battery.battery_wh:.1f}Wh")
                logger.info(f"ì˜ˆìƒ ëŸ°íƒ€ì„: {battery.estimated_runtime:.1f}h")
                logger.info(f"ê±´ê°• ìƒíƒœ: {battery.health_status}")
            
            return response.success
        except Exception as e:
            logger.error(f"ë°°í„°ë¦¬ ìƒíƒœ ì‹¤íŒ¨: {e}")
            return False
    
    def run_all_tests(self):
        """ëª¨ë“  í…ŒìŠ¤íŠ¸ ì‹¤í–‰"""
        logger.info("=== KETI Analysis Engine gRPC Client í…ŒìŠ¤íŠ¸ ì‹œì‘ ===")
        
        if not self.connect():
            logger.error("ì„œë²„ ì—°ê²° ì‹¤íŒ¨. í…ŒìŠ¤íŠ¸ë¥¼ ì¤‘ë‹¨í•©ë‹ˆë‹¤.")
            return
        
        # í…ŒìŠ¤íŠ¸ ì‹œë‚˜ë¦¬ì˜¤
        tests = [
            ("í„°í‹€ë´‡ ìƒì„±", lambda: self.test_create_turtlebot()),
            ("ë””ë°”ì´ìŠ¤ ìƒíƒœ ì¡°íšŒ", lambda: self.test_get_device_status()),
            ("ì‹¤ì œ í„°í‹€ë´‡ InfluxDB ì—…ë°ì´íŠ¸", lambda: self.test_update_from_influx()),
            ("ğŸ¯ MALE Mission ë¶„ì„", lambda: self.test_male_mission_analysis()),
            ("ğŸ“ ì •í™•ë„ ë¶„ì„", lambda: self.test_accuracy_analysis()),
            ("âš¡ ì§€ì—°ì‹œê°„ ë¶„ì„", lambda: self.test_latency_analysis()),
            ("ğŸ† ë””ë°”ì´ìŠ¤ ì¢…í•© ì ìˆ˜", lambda: self.test_device_score()),
            ("ğŸ”‹ ë°°í„°ë¦¬ ìƒíƒœ", lambda: self.test_battery_status()),
            ("ë””ë°”ì´ìŠ¤ ë¶„ì„", lambda: self.test_analyze_device()),
            ("ëª¨ë“  ë””ë°”ì´ìŠ¤ ì¡°íšŒ", lambda: self.test_get_all_devices()),
            ("í”Œë¦¿ ë¶„ì„", lambda: self.test_get_fleet_analysis()),
        ]
        
        success_count = 0
        for test_name, test_func in tests:
            logger.info(f"\n--- {test_name} í…ŒìŠ¤íŠ¸ ---")
            try:
                if test_func():
                    success_count += 1
                    logger.info(f"âœ… {test_name} ì„±ê³µ")
                else:
                    logger.warning(f"âŒ {test_name} ì‹¤íŒ¨")
            except Exception as e:
                logger.error(f"âŒ {test_name} ì˜¤ë¥˜: {e}")
            
            time.sleep(1)  # í…ŒìŠ¤íŠ¸ ê°„ ê°„ê²©
        
        logger.info(f"\n=== í…ŒìŠ¤íŠ¸ ì™„ë£Œ: {success_count}/{len(tests)} ì„±ê³µ ===")
    
    def close(self):
        """ì—°ê²° ì¢…ë£Œ"""
        if self.channel:
            self.channel.close()
            logger.info("ì—°ê²° ì¢…ë£Œ")

def main():
    client = AnalysisClient()
    try:
        client.run_all_tests()
    finally:
        client.close()

if __name__ == "__main__":
    main()